File: lib/src/type_on_call_observer.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:flutter_bridge/src/flutter_method_call_handler.dart';
import 'package:flutter_bridge/src/json_message_codec.dart';

///   author : liuduo
///   e-mail : liuduo@gyenno.com
///   time   : 2022/08/01
///   desc   : 带泛型的OnCall
///   version: 1.0
abstract class TypeOnCallObserver<A, R> implements OnCallObserver {
  @override
  onCall(data) {
    if (data is JsonString) {
      return onCallOfType(fromJsom(json.decode(data.jsonString)));
    }
    return onCallOfType(data as A);
  }

  R onCallOfType(A data);

  A fromJsom(Map<String, dynamic> json) => throw Exception("TypeOnCallObserver 类型A为对象时需要重写此方法");

}

File: android/src/main/java/com/dodo/flutterbridge/TypeOnCallObserver.kt
package com.dodo.flutterbridge

/**
 *     author : liuduo
 *     e-mail : liuduo@gyenno.com
 *     time   : 2022/07/29
 *     desc   : 带泛型的OnCall
 *     version: 1.0
 */
interface TypeOnCallObserver<A, R> : OnCallObserver {
    override fun onCall(data: Any?): Any? {

        if (data is JsonMessageCodec.JsonString) {
            return onCallOfType(fromJson(data))
        }
        return onCallOfType(data as A)
    }

    fun onCallOfType(data: A): R

    fun fromJson(data :JsonMessageCodec.JsonString):A
}